crates/accelerate/src/permutation.rs

```
@@ -0,0 +1,115 @@
+    let n = pattern.len();
+    let mut seen: Vec<bool> = vec![false; n];
+
+    for &x in pattern {
+        if x < 0 {
+            return Err(PyValueError::new_err(
+                "Invalid permutation: input contains a negative number.",
+            ));
+        }
+
+        if x as usize >= n {
+            return Err(PyValueError::new_err(format!(
+                "Invalid permutation: input has length {} and contains {}.",
+                n, x
+            )));
+        }
+
+        if seen[x as usize] {
+            return Err(PyValueError::new_err(format!(
+                "Invalid permutation: input contains {} more than once.",
+                x
+            )));
+        }
+
+        seen[x as usize] = true;
+    }
+
+    Ok(())
+}
+
+fn invert(pattern: &ArrayView1<i64>) -> Array1<usize> {
+    let mut inverse: Array1<usize> = Array1::zeros(pattern.len());
+    pattern.iter().enumerate().for_each(|(ii, &jj)| {
+        inverse[jj as usize] = ii;
+    });
+    inverse
+}
+
+fn get_ordered_swap(pattern: &ArrayView1<i64>) -> Vec<(i64, i64)> {
+    let mut permutation: Vec<usize> = pattern.iter().map(|&x| x as usize).collect();
+    let mut index_map = invert(pattern);
+
+    let n = permutation.len();
+    let mut swaps: Vec<(i64, i64)> = Vec::with_capacity(n);
+    for ii in 0..n {
+        let val = permutation[ii];
+        if val == ii {
+            continue;
+        }
+        let jj = index_map[ii];
+        swaps.push((ii as i64, jj as i64));
+        (permutation[ii], permutation[jj]) = (permutation[jj], permutation[ii]);
+        index_map[val] = jj;
+        index_map[ii] = ii;
+    }
+
+    swaps[..].reverse();
+    swaps
+}
+
+/// Finds inverse of a permutation pattern.
+#[pyfunction]
+#[pyo3(signature = (pattern))]
+fn _inverse_pattern(py: Python, pattern: PyArrayLike1<i64, AllowTypeChange>) -> PyResult<PyObject> {
+    let view = pattern.as_array();
+    validate_permutation(&view)?;
```

I'm not sure this is the best place to validate the permutation (same for the ordered swap), as (1) it's a private utility function without detailed docs and (2) explicit verification is probably done at a different level, if required. It's not super important but it would be slightly faster and the previous implementation also didn't have it 🙂

---

I could go either way on this. It has some overhead (about 10% of the wall-clock time for `_get_ordered_swap` on my machine). On the other hand, if the input is invalid I would rather raise a Python exception than have something go wrong in the Rust layer.

---

Being a non-user-facing, private method I'd say performance is more important and the developer should make sure the input makes sense. We could expose your validation function e.g. as `_validate_permutation` to Python and use it in the appropriate places.

---

Sounds fine, I moved it into a separate pyfunction.